packages/teleport/src/Discover/LoginTrait/LoginTrait.tsx

```diff
@@ -0,0 +1,169 @@
+                  css={`
+                    border: 1px solid
+                      ${props => props.theme.colors.primary.light};
+                  `}
+                >
+                  <input
+                    type="checkbox"
+                    name={login}
+                    checked={loginMap[login]}
+                    css={`
+                      margin-right: 10px;
+                      accent-color: ${props =>
+                        props.theme.colors.secondary.main};
+                      &:hover {
+                        cursor: pointer;
+                      }
+                    `}
+                    onChange={() => toggleLoginSelect(login)}
+                  />
+                  <label
+                    htmlFor={login}
+                    css={`
+                      width: 250px;
+                      overflow: hidden;
+                      text-overflow: ellipsis;
+                    `}
+                  >
+                    {login}
+                  </label>
+                </Flex>
+              );
+            })}
+            {showInputBox && (
+              <Flex alignItems="end" mt={3}>
+                <Validation>
+                  <FieldInput
+                    placeholder="name"
+                    autoFocus
+                    width="200px"
+                    value={newLogin}
+                    type="text"
+                    onChange={e => setNewLogin(e.target.value.trim())}
+                    mr={3}
+                    mb={0}
+                  />
+                  <ButtonPrimary
+                    size="small"
+                    mb={2}
+                    disabled={newLogin.length === 0}
+                    onClick={() => {
+                      addLogin(newLogin);
+                      setNewLogin('');
+                      setShowInputBox(false);
+                    }}
+                  >
+                    Add
+                  </ButtonPrimary>
+                </Validation>
+              </Flex>
+            )}
+            {!showInputBox && (
+              <ButtonText
+                mt={2}
+                onClick={() => setShowInputBox(true)}
+                css={`
+                  line-height: normal;
+                  padding-left: 4px;
+                `}
+              >
+                <Icons.Add
+                  css={`
+                    font-weight: bold;
+                    letter-spacing: 4px;
+                    &:after {
+                      content: ' ';
+                    }
+                  `}
+                />
+                Add New Principle
+              </ButtonText>
+            )}
+          </Box>
+          <ButtonPrimary
+            width="165px"
+            onClick={() => {
+              let checkedLogins = Object.keys(loginMap).filter(login =>
+                loginMap[login] ? login : false
+              );
+              nextStep(checkedLogins);
+            }}
```

Comment:
I'd move this into a callback using `React.useCallback` -

```typescript
const handleClick = React.useCallback(() => {
  // etc...
}, [loginMap, nextStep]);
```

Reply:
Also, if you use a map instead of a record here, you can do something like

```
const checkedLogins = [...loginMap.entries()]
  .filter(([, value]) => value)
  .map(([key]) => key)
```

It's a bit nicer to iterate on a map than the keys of an object

Reply:
> I'd move this into a callback using React.useCallback -

hm.. i don't agree in using `useCallback` here, the logic is not intensive enough to warrant a optimizing hook and i don't expect the list of user names to be gigantic. 

i'd be curious in a third opinion though b/c when to use these optimizing hooks is confusing and i typically reach for them (rarely) when "is it computationally intensive? or can the performance problem be measured or is visibly problematic?". Opinions? @hatched @ravicious @gzdunek

Reply:
I agree that the performance of it is barely noticeable, I just prefer to keep the business logic for a handler out of the JSX code.

Note that not using inline functions is quite a popular approach, there's an [ESLint rule](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/jsx-no-bind.md) for it too.

Not gonna push on this as it comes down to personal preference really

Reply:
I typically try and keep function definitions out of the `render` section of a react component for legibility but in this specific case it's small enough I'm not super concerned if you felt strongly about leaving it here.

Now as far as a map is concerned....generally you don't want to use user defined input as keys as you could end up with some unintended consequences but using a `Map` would give you some protections against this.

For this specific case however....outside of maintaining UI state I don't see the `checked` logic being used for anything before the 'form' is to be submitted. So I would let the browser handle the state. By this I mean wrap the inputs in a form element with a `ref`. Then use "uncontrolled" components with a defined `name` for the checkboxes and only on "next" grab the ref and parse the values into an array of values `['root', 'ubuntu', ...]` and store that.

Reply:
> I would let the browser handle the state. By this I mean wrap the inputs in a form element with a ref. Then use "uncontrolled" components with a defined name for the checkboxes and only on "next" grab the ref and parse the values into an array of values ['root', 'ubuntu', ...] and store that.

@hatched that was a brilliant suggestion, simplified a lot of things

Reply:
Haha thanks! Happy to help 😁